### PR TITLE
Elide unused import errors for failed use statements

### LINKED ISSUE #48244: "unused imports" lints run too eagerly
Ever since @alexcrichton refactored lints (which was awesome!) to print eagerly, the unused import lints have gotten pretty annoying. Example output from a recent compilation session:

```
error[E0432]: unresolved import `rustc::util::CellUsizeExt`
  --> src/librustc_traits/normalize_projection_ty.rs:19:5
   |
19 | use rustc::util::CellUsizeExt;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^ no `CellUsizeExt` in `util`

warning: unused import: `rustc::util::CellUsizeExt`
  --> src/librustc_traits/normalize_projection_ty.rs:19:5
   |
19 | use rustc::util::CellUsizeExt;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: #[warn(unused_imports)] on by default
   ^^^^^^^^^^^^^^^^^^^^^^^^^
```

Note that I got an error about the import not being valid -- and then a lint that my invalid path is not used! Here is a [similar problem](https://play.rust-lang.org/?gist=b7f647d732e04dd9cb9c875b6a5bd92e&version=stable), in a standalone test case:

```rust
use std::io::Write;

fn main() {
    let x = File::open(22); // deliberate missing import
    x.write(22);
}
```

As of now, I get this output (in stable, but also nightly):

```
   Compiling playground v0.0.1 (file:///playground)
error[E0433]: failed to resolve. Use of undeclared type or module `File`
 --> src/main.rs:4:13
  |
4 |     let x = File::open("foo");
  |             ^^^^ Use of undeclared type or module `File`

warning: unused import: `std::io::Write`
 --> src/main.rs:1:5
  |
1 | use std::io::Write;
  |     ^^^^^^^^^^^^^^
  |
  = note: #[warn(unused_imports)] on by default
```

I have also frequently seen it happen that the unused import lint comes **first**, which is extra annoying.

### PR DESCRIPTION
**Pull request by estebank**
Fix #48244

### CODE REVIEW

File: src/librustc_resolve/check_unused.rs [hunk 1]
```diff
@@ -305,6 +307,31 @@ pub fn check_crate(resolver: &mut Resolver<'_>, krate: &ast::Crate) {
...
         let len = spans.len();
         spans.sort();
+        for sp in &spans {
+            debug!(". {:?}", sp);
+            // Do not emit unused warnings about failed use statements (#48244)
+            for &crate::PrivacyError(dedup_span, _, _) in &visitor.privacy_errors {
+                debug!("- {:?}", dedup_span);
+                if sp.overlaps(dedup_span) {
+                    debug!("break");
+                    break 'unused_imports;
```
Comment (cramertj) [inline src/librustc_resolve/check_unused.rs]: Wouldn't this get rid of *all* unused import errors in the case that *any* import failed? (that is to say, I'd have expected this to be `continue 'unused_imports`)

Comment (estebank) [reply src/librustc_resolve/check_unused.rs]: Indeed. I made a few other changes, including replacing `break` for `continue`. (This was leftover from wip.)

**Comment by estebank** [comment]
If we don't want to pay the price of iterating over all the spans, we can simply silence _all_ unused import lints if any import errors have been found, but I slightly prefer this more targeted approach.

**Comment by petrochenkov** [comment]
Unresolved imports are resolved to `Def::Err` during "import finalization".
Every time we perform such a resolution for an import we need to mark it as used.
I don't quite remember where it's done exactly, will look today.

**Comment by petrochenkov** [comment]
Superseded by https://github.com/rust-lang/rust/pull/60359

**Pull request closed without merging**